File: bme6315/LRparamEst2.m
% LRparamEst2.m
% parameter estimation for 2 fitted parameters

function output = LRparamEst2
%% Parameter assignments
L = 0.9;
kon = 1.1;    % fitting these, put best guess
koff = 0.1; % fitting these, put best guess
% params = [L,kon,koff];

% Simulation parameters
y0 = [1; 0];    % initial conditions
tspan = [0 6];
simOptions = [];

% Load experimental data (synthetic, using L = 1, kon = 1, koff = 0.1)
data = [0:6; 0,0.58,0.79,0.93,0.88,0.94,0.80]';  % 1st column: times; 2nd column: LR measurements

paramsFixed = [koff];     % designate parameters that will not be estimated
lowerbnd = [];              % optional lower bound on parameter estimate
upperbnd = [];              % optional upper bound on parameter estimate
params0 = [L,kon];            % initial guesses of parameter estimate
% params0 = (upperbnd-lowerbnd).*rand(size(params0))+lowerbnd;    % use random initial guesses (overwrites above line)

optimOptions = optimset('Display','iter');
objFcn = @(params0) objectiveFcn(params0,paramsFixed,tspan,y0,data,simOptions);
[paramsEst,resnorm,residual,exitflag,output,lambda,Jacobian]=lsqnonlin(objFcn,params0,lowerbnd,upperbnd,optimOptions);
disp(['Algorithm: ',output.algorithm]);
disp(['Exit flag: ',num2str(exitflag),', resnorm = ',num2str(resnorm)]);
paramsVariance = resnorm*inv(Jacobian'*Jacobian)/length(data(:,1));  % estimates the parameter estimate covariance using the Cramer-Rao inequality
paramsStd = sqrt(diag(paramsVariance));   % standard deviations for parameter estimates
paramsCI = 1.96*paramsStd;                % mean +/- 1.96*stdev in 95% confidence interval
output = [paramsEst;paramsCI']

%% Run test of parameter estimation
params = [paramsEst(1),paramsEst(2),koff];
[tSingle,ySingle] = ode23(@LRodeFunc,tspan,y0,simOptions,params);
LRsingle = ySingle(:,2);
figure(1);
subplot(1,3,1);
plot(tSingle,LRsingle,data(:,1),data(:,2),'o');
xlabel('Time (sec)'); ylabel('LR (\muM)');
legend('model','data');
subplot(1,3,2);
barweb([[1.0; 1.0],paramsEst'],[[0; 0], paramsStd],[],{'L','kon'}); % note need 'barweb' from MatlabCentral
axis([0.5 2.5 0 1.5])
legend('Original','Estimated'); ylabel('Estimated parameters');

%% Plot entire SSE surface to verify that paramEst is in fact a minimum
LEst = paramsEst(1);
konEst = paramsEst(2);
SSEest = sum(objFcn(paramsEst).^2);
param1Range = [0.5:.05:2];
param2Range = [0.5:.05:2];
for i = 1:length(param1Range)
    for j = 1:length(param2Range)
        param = [param1Range(i),param2Range(j)];
        SSE(i,j) = sum(objFcn(param).^2);
    end
end
subplot(1,3,3);
imagesc(param1Range,param2Range,SSE); 
xlabel('L'); ylabel('kon'); 
c = colorbar;
c.Label.String = 'SSE';

function error = objectiveFcn(paramsEst,paramsFixed,tspan,y0,data,simOptions)
% objective function for performing least squares minimization
% This error function is weighted by the std dev of the experimental
% measurements, which affects confidence intervals and relative weighting 
% of multi-objective function estimates.
L = paramsEst(1);
kon = paramsEst(2);
koff = paramsFixed(1);
params = [L,kon,koff];

[tSim,ySim] = ode23(@LRodeFunc,tspan,y0,simOptions,params);   % run the model for a given sample of paramsEst
ySimInterp = interp1(tSim,ySim,data(:,1)); % resample simulated data before performing subtraction
stdevExp = .05; % standard deviation of experimental data used in weights (set to 1 by default)
error = (1/stdevExp).*(ySimInterp(:,2) - data(:,end));     % note that LSQNONLIN computes the sum of squares for this automatically

File: bme6315/python/LRparamEst2.py
# LRparamEst.py
# Parameter estimation for a ligand-receptor model
# 10/1/2021 by JS
import numpy as np
from scipy import optimize, integrate, interpolate, linalg
import matplotlib.pyplot as plt

# Parameter assignments
L = 1
kon = 1
koff = 0.1
params = [L,kon,koff]

def ODEfunc(t,y,params):
    L,k1f,k1r = params
    R, LR = y    
    react1 = k1f*L*R - k1r*LR       # [uM/s] L+R<->LR    
    dR = -react1            # [uM/s] free receptor
    dLR = react1            # [uM/s] ligand-receptor complex    
    return [dR,dLR]

# Simulation parameters
y0 = [1,0]
tspan = [0,6]
paramsFixed = koff
p0 = [0.9, 1.1]

# Load experimental data
xdata = np.arange(7)
ydata = [0,0.58,0.79,0.93,0.88,0.94,0.80]

# Run parameter estimation
def objectiveFunction(paramsEst,paramsFixed,tspan,xdata,ydata):
    koff = paramsFixed
    L, kon = paramsEst
    params = [L,kon,koff]
    sol = integrate.solve_ivp(ODEfunc,tspan,y0,args=(params,))
    finterp = interpolate.interp1d(sol.t,sol.y[1,:]) # create an interpolation function
    error = ydata - finterp(xdata)       #  compute error at resampled locations
    return error

result = optimize.least_squares(objectiveFunction, p0, args=(paramsFixed,tspan,xdata,ydata))
paramsEst = result.x

# Run test of parameter estimation, shows a good fit
params = [paramsEst[0],paramsEst[1],koff]
sol = integrate.solve_ivp(ODEfunc,tspan,y0,args=(params,))
plt.plot(xdata,ydata,'o',sol.t,sol.y[1,:].T,'-')
plt.show()

# generate parameter ensemble
# note in this particular example the covariance is huge. If the parameters had
# been more identifiable it would be much smaller.
covariance = result.cost*linalg.inv((np.matmul(result.jac.T,result.jac)))/len(ydata)
n = 10;
paramEnsemble = np.random.multivariate_normal(paramsEst,covariance,n)
paramsStd = np.std(paramEnsemble,axis=0)

# plot parameter estimates
plt.bar(['L','kon'],paramsEst)
plt.errorbar(['L','kon'],paramsEst,paramsStd)
plt.show()